### PR TITLE
Fix inefficient label mapping in direct color mode (10-20x speedup)

### PR DESCRIPTION
**Pull request by ksofiyuk**
# Description

If you specify a custom color set in the Label layer, the brush becomes extremely laggy in high-resolution images. After some examination, I discovered that issue is caused by the inefficient label mapping in the direct color mode:
https://github.com/napari/napari/blob/a79537cac7c2a595c4fca41b85f2c7b4a44d0a61/napari/layers/labels/labels.py#L927-L935

It calls `np.unique` of the whole label map every brush update! This is a very slow operation for large arrays. As you can see in the benchmark results below, it takes 1.5s for a single `_raw_to_displayed` call on 4096x4096 images.

Luckily, in most use cases the use of `np.unique` is a total overkill here and it is possible to achieve the same functionality just by finding the minimum and maximum values of an array. I implemented a new algorithm, which is 10-20x faster on large images. To be on the safe side, I kept the old approach for the cases when `max_label_id - min_label_id >= 1024`, it will handle the situations when someone decides to load labels with very large ids (e.g. label_id=123456).

I also added a new benchmark `Labels2DColorDirectSuite` that tests the Label layer in the direct color mode.

## Benchmark results:
The results for the Label layer in the `AUTO` mode (for comparison):
```
[100.00%] ··· benchmark_labels_layer.Labels2DSuite.time_raw_to_displayed                                                                                                                          ok
[100.00%] ··· ======== =============
               param1
              -------- -------------
                 16     3.84±0.02μs
                 32     5.89±0.03μs
                 64     14.4±0.04μs
                128     47.6±0.09μs
                256       183±1μs
                512     1.02±0.01ms
                1024    4.05±0.02ms
                2048     15.8±0.2ms
                4096      63.7±2ms
              ======== =============
```

The results for the Label layer in the `DIRECT` mode (old algorithm):
```
[100.00%] ··· benchmark_labels_layer.Labels2DColorDirectSuite.time_raw_to_displayed                                                                                                               ok
[100.00%] ··· ======== ============
               param1
              -------- ------------
                 16     36.5±0.2μs
                 32      47.2±1μs
                 64      150±2μs
                128      534±2μs
                256     2.31±0.2ms
                512     11.8±0.1ms
                1024    61.8±0.7ms
                2048     312±2ms
                4096    1.50±0.02s
              ======== ============
```

The results for the Label layer in the `DIRECT` mode (new algorithm):
```
[100.00%] ··· benchmark_labels_layer.Labels2DColorDirectSuite.time_raw_to_displayed                                                                                                               ok
[100.00%] ··· ======== =============
               param1
              -------- -------------
                 16       17.1±1μs
                 32     20.0±0.04μs
                 64      35.1±0.3μs
                128      93.8±0.2μs
                256       330±2μs
                512     1.63±0.01ms
                1024    6.40±0.02ms
                2048     25.1±0.1ms
                4096     101±0.5ms
              ======== =============
```

As you can see, it became 15x faster on 4096x4096 images, achieving almost the same performance as in the `AUTO` mode.

## Further optimization

It is just a quick fix to mitigate the issue. However, I think it is very inefficient to compute the label mapping from scratch on every brush update. At least the proposed approach can be further optimized in the following way:

1. Recompute `min_label_id` and `max_label_id` from scratch only when someone tries to set the data in a layer.
2. When someone tries to change the color id of brush, update the `min_label_id` and `max_label_id` with a new color id accordingly.

This should be enough to always maintain the valid `min/max` of label ids without the need of recomputing them on every update.

If we want to be ultimately efficient, the color map should be computed only for the part of an image that is visible in the current camera view. Or only recompute the color map near the local region around a recent brush trajectory as during drawing in large images 99% of image area usually is not affected and should not be updated. It should allow to handle extremely large images (e.g. 10000x10000). 

## Type of change
- [X] Bug-fix (non-breaking change which fixes an issue)

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by Czaki** [comment]
Our current target for solving this problem is to avoid global remapping of labels on CPU in #3308 but I'm not sure when linked PR will be finished. 

We will try to investigate of the status of #3308 and if the prognosis of the finish will be not optimistic I will be happy to at least partially solve this performance problem by merging this PR. 

If you have time You could also take a look at the linked PR and maybe help to finish it.

**Comment by Czaki** [comment]
one idea (the changes will be bigger) 

We do not need to know the real set of labels. We only need to have a guarantee that all labels from layers are present in this set (so even if we erase some label, then its present is not a problem). 

So the call of unique could be delayed and done for example 1s after finishing the drawing (with updating the set of every variable used in the drawing). 

It will be more complex and take a long time with deeper dive in napari code, but it also will be useful for solve problems mentioned in #3308 and speedup code even more (as avoid calculate min and max on every refresh).

**Comment by jni** [comment]
@ksofiyuk How does your benchmark suite fare on #3308? I think that PR makes this one unnecessary, and it's very important (as you'll see in the discussion there, it fixes a *heap* of issues with direct color mode), so I think we should focus on that one...

**Comment by ksofiyuk** [comment]
@jni I'll test that PR later. Anyway, I propose to merge this PR if you don't have any objections regarding the code. It is a non-breaking change that doesn't change any behaviour - a pure optimization. 

I did it because without it napari is unusable on large images with custom colors, it is not a matter about optimization for the sake of optimization. Just try it yourself, open a large image (I use 4096x3000) and try to use the brush. You should consider it as a bugfix or a necessary optimization.

I can't see any conflicts with the mentioned PR because if it resolves the issue, then when it's ready this code will be unnecessary and will be replaced.

**Comment by brisvag** [comment]
> To be on the safe side, I kept the old approach for the cases when `max_label_id - min_label_id >= 1024`, it will handle the situations when someone decides to load labels with very large ids (e.g. label_id=123456).

To be clear, this is to handle better cases where labels are sparsely distributed, like `[0, 1, 12341234, 893475892374]`?

> I can't see any conflicts with the mentioned PR because if it resolves the issue, then when it's ready this code will be unnecessary and will be replaced.

I agree with @ksofiyuk, that PR has been sitting for a long time, who knows how much longer, and this one brings very little disruption. Tests are passing already, which is a good sign. I'm not super familiar with the logic here though, so I'm not confident that it won't break some corner cases. Hopefully @jni can help more here.

File: napari/layers/labels/labels.py
Comment (alisterburt) [inline napari/layers/labels/labels.py:928]:
```suggestion
            max_label_id = raw_modified.max()
            n_unique_labels_upper_bound = max_label_id - min_label_id
```

File: napari/layers/labels/labels.py
Comment (alisterburt) [inline napari/layers/labels/labels.py]:
```suggestion
            if n_unique_labels_upper_bound < 1024:
```

**Review by alisterburt — APPROVED** [review]
I'm also fine with this PR going in whilst waiting for #3308 - I know that's getting closer but this will be easy to remove once #3308 is ready and the performance gains here are high value

I defined the upper bound on the number of unique labels explicitly which I think makes the code a little easier to parse at first glance

**Comment by alisterburt** [comment]
thanks again @ksofiyuk - you're on fire! 🔥

File: napari/layers/labels/labels.py
Comment (alisterburt) [inline napari/layers/labels/labels.py]:
```suggestion
            upper_bound_n_unique_labels = max_label_id - min_label_id
```

File: napari/layers/labels/labels.py
Comment (alisterburt) [inline napari/layers/labels/labels.py]:
```suggestion
            if upper_bound_n_unique_labels < 1024:
```

**Comment by ksofiyuk** [comment]
@alisterburt thanks for the improving of readability.

1. I updated the threshold for `upper_bound_n_unique_labels` to 65536, as it can handle even that range of labels without almost any performance drop. Actually, we can safely increase it even more, but I think 65536 should cover 99.999% of possible cases.
2. I also updated the benchmark to incorporate a larger range of generated labels (from -10000 to 10000).

In this benchmark with sparse labels from -10000 to 10000, the proposed method shows even much more significant speedup (up to 40x).

The timing of the old algorithm on a new benchmark (provide only the two biggest sizes for clarity):
```
                2048      521±10ms  
                4096     2.63±0.02s 
```

The new algorithm:
```
                2048     17.3±0.2ms 
                4096     62.3±0.8ms 
```

It is 42x speedup!

I tested it on a bit different hardware, so the numbers don't exactly match the results in the first message.

**Comment by ksofiyuk** [comment]
> To be clear, this is to handle better cases where labels are sparsely distributed, like [0, 1, 12341234, 893475892374]?

@brisvag, no, on the contrary, it fixes the cases when labels are not super sparsely distributed (I can say it covers 99.999% of real world cases). With the last update, you will gain significant speedup from this PR, if your labels can be localized in any continuous interval which length is less than 65536 elements. Here are some examples, if your labels are:
```
[0, 1, 2, 3]   # you will get significant boost here
[-1000, 0, 1, 1000]  # you will also get significant boost here as it falls inside the interval of length 65536
[100000, 100100, 100200, 100300]  # again it falls inside the interval, the new algorithm will be used
[-1000000, 1000000]  # they cannot be localized in a continuous of length 65536, the old method with `np.unique` will be used.
```

**Comment by jni** [comment]
Super cool! Yeah, sorry, I didn't mean to imply we shouldn't merge this! I actually think the other one is closer than you're implying @brisvag but I totally get the point that this is already passing and a huge speedup so I'm all for merging! The other factor is that I think using direct mode, sooner rather than later @ksofiyuk will run into some of the color mismatch issues fixed by #3308. But we can cross that bridge when we get there!

**Review by jni — APPROVED** [review]

**Comment by brisvag** [comment]
Sorry, with "this" I meant "keeping the old approach for some cases" :P So I'm on board :+1:

**Comment by brisvag** [comment]
> I actually think the other one is closer than you're implying @brisvag 

Nice, looking forward to it :)

**Comment by jni** [comment]
Just pushed a commit trying to make sure all the new lines are covered. This should be good to merge if that passes, deepest apologies if it doesn't 🤣 (it passed locally)

**Comment by ksofiyuk** [comment]
> Just pushed a commit trying to make sure all the new lines are covered. This should be good to merge if that passes, deepest apologies if it doesn't rofl (it passed locally)

I'm sorry, but this test is not representative for this code :smile:, because the range of labels is going beyond 65536 (2 ** 16), so the previous version of the mapping will be invoked.

**Comment by jni** [comment]
@ksofiyuk that's the point — the old tests didn't have a branch, but they were "shuttled" to the new code — if you look at the previous codecov comment, it was the *old* code that was no longer covered by tests. Hence the new test.

Anyway, new npe2 release broke our typing action. CC @tlambert03

**Comment by ksofiyuk** [comment]
> that's the point — the old tests didn't have a branch, but they were "shuttled" to the new code — if you look at the previous codecov comment, it was the _old_ code that was no longer covered by tests. Hence the new test.

@jni I realized it after I posted it, so I deleted my previous comment, but you managed to read it and reply before it :smile: But thanks for the explanation.

**Comment by ksofiyuk** [comment]
> How does your benchmark suite fare on https://github.com/napari/napari/pull/3308?

@jni I've tested PR #3308 and it indeed has a much more efficient implementation of the label rendering.

This PR:
```
1024    9.17±0.06ms
2048     27.7±0.1ms
4096      103±3ms
8192      409±10ms
16384     1.70±0.02s
```

PR #3308:
```
1024    1.17±0.01ms
2048    4.50±0.06ms
4096     17.5±0.3ms
8192      71.1±8ms
16384      296±9ms
```

So it is around 5-6x faster than the current implementation on large images. However, you can see that even the GPU implementation has exactly the same pattern of growth, it is just faster because of utilizing better hardware.

I believe that the label rendering can be significantly improved by some clever algorithmic optimizations which would allow to have a near constant time complexity of rendering on a brush update without the dependency on the resolution of an image.

**Comment by Czaki** [comment]
> I believe that the label rendering can be significantly improved by some clever algorithmic optimizations which would allow to have a near constant time complexity of rendering on a brush update without the dependency on the resolution of an image.

One of the core problems here is that we transfer the whole layer array to GPU after paint. 
So bigger array means more linear operations. 

I do not have time to investigate how it will be hard to transfer only the updated part of the image. 

But this plus maybe caching some properties should lead to better performance

**Comment by ksofiyuk** [comment]
> One of the core problems here is that we transfer the whole layer array to GPU after paint.
> So bigger array means more linear operations.

In theory, it shouldn't be very difficult to send to GPU for updating only patches of an image. But I'm also not sure how difficult it would be to implement in napari (I haven't looked into the code of PR #3308).

It does not look very difficult to implement local updates (only around the recent brush movement + limit the updating only to the part of the image that can be seen) for the CPU version. And it should allow to handle images of any resolution without performance degradation for 99,9% of typical use cases. And it would beat the GPU version on really high resolution images if the same thing wasn't implemented for it.

**Comment by alisterburt** [comment]
> It does not look very difficult to implement local updates (only around the recent brush movement + limit the updating only to the part of the image that can be seen) for the CPU version. And it should allow to handle images of any resolution without performance degradation for 99,9% of typical use cases. And it would beat the GPU version on really high resolution images if the same thing wasn't implemented for it.

This would be cool, here is a previous implementation of partial updates in the volume visual which may provide some clues ☺️
https://github.com/napari/napari/pull/1979

**Comment by Czaki** [comment]
Currently we also send data to GPU for render. I try to create graph:

obraz

For #3308 it will looks like (I grayed out current steeps):


obraz


Casting to float32 is much faster than the current steep but still linear in data size. And in both scenarios, there is a transfer to GPU steep. 


(On both graphs I highlight only most important steeps, it is not a full documentation)

**Comment by ksofiyuk** [comment]
@Czaki Thank you for this graph. Based on this graph, I'd say the only correct way to make it very efficient is to implement sending local updates to GPU instead of a whole image and label map.

If the above is not done, there is no reason in further CPU optimization as the transferring data from CPU to GPU will always be the bottleneck.

**Comment by alisterburt** [comment]
This is marked as ready for merge and tests are passing but something is up with the [typing test](https://github.com/napari/napari/actions/runs/4700373164/jobs/8335070666?pr=5723) - I'm not familiar with this and the error appears to be in files unaffected by this PR. I will merge in a few hours unless anyone objects

**Comment by alisterburt** [comment]
just seen that #5732 includes this

**Comment by Czaki** [comment]
@alisterburt Typing fix is here #5727

**Comment by Czaki** [comment]
@jni Labels and milestone

**Comment by jni** [comment]
@Czaki should we make a 0.4.18 milestone for cherry-picking?

**Comment by alisterburt** [comment]
+1 on the 0.4.18 milestone

**Comment by Czaki** [comment]
@jni yes

**Pull request merged**
